x-pack/filebeat/input/s3/input.go

```diff
@@ -0,0 +1,507 @@
+	waitTimeSecond int64 = 10
+
+	errOutletClosed = errors.New("input outlet closed")
+)
+
+func init() {
+	err := input.Register(inputName, NewInput)
+	if err != nil {
+		panic(err)
+	}
+}
+
+// Input is a input for s3
+type Input struct {
+	outlet     channel.Outleter // Output of received s3 logs.
+	config     config
+	awsConfig  awssdk.Config
+	logger     *logp.Logger
+	close      chan struct{}
+	workerOnce sync.Once // Guarantees that the worker goroutine is only started once.
+	context    *channelContext
+	workerWg   sync.WaitGroup // Waits on s3 worker goroutine.
+}
+
+type s3Info struct {
+	name   string
+	key    string
+	region string
+	arn    string
+}
+
+type bucket struct {
+	Name string `json:"name"`
+	Arn  string `json:"arn"`
+}
+
+type object struct {
+	Key string `json:"key"`
+}
+
+type s3BucketOjbect struct {
+	bucket `json:"bucket"`
+	object `json:"object"`
+}
+
+type sqsMessage struct {
+	Records []struct {
+		EventSource string         `json:"eventSource"`
+		AwsRegion   string         `json:"awsRegion"`
+		EventName   string         `json:"eventName"`
+		S3          s3BucketOjbect `json:"s3"`
+	} `json:"Records"`
+}
+
+type s3Context struct {
+	mux  sync.Mutex
+	refs int
+	err  error // first error witnessed or multi error
+	errC chan error
+}
+
+// channelContext implements context.Context by wrapping a channel
+type channelContext struct {
+	done <-chan struct{}
+}
+
+func (c *channelContext) Deadline() (time.Time, bool) { return time.Time{}, false }
+func (c *channelContext) Done() <-chan struct{}       { return c.done }
+func (c *channelContext) Err() error {
+	select {
+	case <-c.done:
+		return context.Canceled
+	default:
+		return nil
+	}
+}
+func (c *channelContext) Value(key interface{}) interface{} { return nil }
+
+// NewInput creates a new s3 input
+func NewInput(cfg *common.Config, connector channel.Connector, context input.Context) (input.Input, error) {
+	cfgwarn.Beta("s3 input type is used")
+	logger := logp.NewLogger(inputName)
+
+	config := defaultConfig()
+	if err := cfg.Unpack(&config); err != nil {
+		return nil, errors.Wrap(err, "failed unpacking config")
+	}
+
+	out, err := connector.ConnectWith(cfg, beat.ClientConfig{
+		Processing: beat.ProcessingConfig{
+			DynamicFields: context.DynamicFields,
+		},
+		ACKEvents: func(privates []interface{}) {
+			for _, private := range privates {
+				if s3Context, ok := private.(*s3Context); ok {
+					s3Context.done()
+				}
+			}
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	awsConfig, err := awscommon.GetAWSCredentials(config.AwsConfig)
+	if err != nil {
+		return nil, errors.Wrap(err, "getAWSCredentials failed")
+	}
+
+	closeChannel := make(chan struct{})
+	p := &Input{
+		outlet:    out,
+		config:    config,
+		awsConfig: awsConfig,
+		logger:    logger,
+		close:     closeChannel,
+		context:   &channelContext{closeChannel},
+	}
+	return p, nil
+}
+
+// Run runs the input
+func (p *Input) Run() {
+	p.workerOnce.Do(func() {
+		p.workerWg.Add(1)
```

Comment:
please add this line right before you call `go p.run`. The `getRegionFromQueueURL` can fails, leaving you with an invalid count on workerWg. The Beat would deadlock on shutdown in this case.